core/src/apps/cardano/address.py

```diff
@@ -1,54 +1,104 @@
-from trezor import log
+from trezor import log, wire
 from trezor.crypto import base58, crc, hashlib
 
+from apps.cardano.protocol_magics import PROTOCOL_MAGICS
 from apps.common import HARDENED, cbor
 from apps.common.seed import remove_ed25519_prefix
 
+if False:
+    from typing import Tuple
+    from trezor.crypto import bip32
+    from apps.cardano import seed
 
-def _encode_address_raw(address_data_encoded):
+PROTOCOL_MAGIC_KEY = 2
+INVALID_ADDRESS = wire.ProcessError("Invalid address")
+
+
+def _encode_address_raw(address_data_encoded: bytes) -> str:
     return base58.encode(
         cbor.encode(
             [cbor.Tagged(24, address_data_encoded), crc.crc32(address_data_encoded)]
         )
     )
 
 
-def derive_address_and_node(keychain, path: list):
+def derive_address_and_node(
+    keychain: seed.Keychain, path: list, protocol_magic: int
+) -> Tuple[str, bip32.HDNode]:
     node = keychain.derive(path)
 
-    address_payload = None
-    address_attributes = {}
+    address_attributes = get_address_attributes(protocol_magic)
 
-    address_root = _get_address_root(node, address_payload)
+    address_root = _get_address_root(node, address_attributes)
     address_type = 0
     address_data = [address_root, address_attributes, address_type]
     address_data_encoded = cbor.encode(address_data)
 
     return (_encode_address_raw(address_data_encoded), node)
 
 
-def is_safe_output_address(address) -> bool:
+def get_address_attributes(protocol_magic: int) -> dict:
+    # protocol magic is included in Byron addresses only on testnets
+    if protocol_magic == PROTOCOL_MAGICS["mainnet"]:
+        address_attributes = {}
+    else:
+        address_attributes = {PROTOCOL_MAGIC_KEY: cbor.encode(protocol_magic)}
+
+    return address_attributes
+
+
+def is_safe_output_address(address: str) -> bool:
     """
     Determines whether it is safe to include the address as-is as
     a tx output, preventing unintended side effects (e.g. CBOR injection)
     """
```

Comment:
ISTM now that you're not using `Raw` anymore, there is no CBOR injection. I am not sure what this check is accomplishing exactly.

In fact the structure of the whole file is unclear. There is `_encode_address_raw`, and `_get_address_data` seems to be its counterpart -- but in that case, it should perform the CRC check, and maybe even validate that the CBOR tag is present? I'm not sure how important that is though.
Also the names should make it clear that these two functions perform the reverse operations. Something like `_wrap_address_raw` and `_unwrap_address_raw`?

And then I'd rename `matches_with_protocol_magic` to `validate_address(address: str, protocol_magic: int)` and make it raise either `INVALID_ADDRESS` if the format is wrong, or "Network mismatch" if the protocol magic is wrong -- so that in `sign_tx` you can just call `validate_address()` without the ifs

Reply:
While you're at this, please remove `Raw` from the CBOR encoder. It's dangerous and now that we don't need it, there is no reason for it to live in the code.

Reply:
We are still using `Raw` so the check is still valid. I agree that the file is unclear, but it's also given by the fact that the Byron address format is complicated. I'm trying to refactor it with your notes in mind.

Reply:
i don't see any instances of `Raw` in the code. Are Shelley addresses using `Raw`?

Reply:
Hm, I see. I was a bit confused. Indeed, output addresses are encoded as bytes now. Shelley addresses are also not using `Raw`. So this check can be removed - I'll confirm this with the team.

Reply:
Done [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/10305274dbc6f66e0e7007c0a7ecbc0fddbc1372). CBOR injection check has been removed. We now need to deconstruct the address only to get to the protocol magic. CRC doesn't need to be checked since that is done by the node.

I've also removed `Raw` [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/0ead85323b19db525f48336936246602f066f8eb) from CBOR. Unfortunately I can't remove the 24 Tag since it's still used when building the address from path.

Reply:
The CRC check should be there anyway, to prevent user error. We shouldn't ask the user to confirm address strings which we can determine are invalid.
Unless there is a valid usecase for generating txes to invalid addresses, of course .)

Reply:
Added [here](https://github.com/trezor/trezor-firmware/pull/1096/commits/ea4a4b36c687f32f9d3385aef083bccfe7631e78). I also added a test case for that.

Reply:
Oh, I need to generate UI fixtures again.